### ISSUE TITLE
Binary files upload/download support

### ISSUE THREAD

**Issue by dudicp** [cometd/cometd#689]
Hi,

I’m currently evaluating cometd for a new project I’m working on and it’s seems that cometd does provide most of the functionalities I’m looking for.
However, one of my requirements is to support upload/download of binary files (large) over bayeux session (web-sockets) and I haven’t found much information regarding this, hence I’m wondering if cometd is really suitable for this requirement or this will be abusing of the cometd library?

Thanks,
David

**Comment by sbordet** [comment]
See #446 and https://docs.cometd.org/current/reference/#_concepts_binary_data (just fixed).

It should not be much work to implement Z85, which is designed to be superior to Base64, and it should be handled transparently for applications via a CometD extension.

**Comment by dudicp** [comment]
Thanks. 
Regarding large files, I assume I need to implement some mechanism that breaks the file to multiple chunks and reconstruct the file on the server side? there is no such built-in implementation in the library?

**Comment by sbordet** [comment]
The HTTP transport does not have any problem uploading messages of any size; however, the WebSocket transport has limitations on the max message size, so in general yes, a mechanism for splitting the file would be needed, and it could probably be integrated in the extension that does Z85.

What exactly your use case ? From JavaScript using `Blob` ? Or in memory using `ArrayBuffer` ?

**Comment by dudicp** [comment]
My requirement probably isn’t the most common use-case for cometd. The requirement is to implement a server to server communication over reverse HTTP connection (such as web-sockets or long polling). Both servers are implemented in java and using the cometd java server & client libraries accordingly.
Also, the requirement is to support both text-based messages and binary-based messages in efficient way (in terms of memory usage, latency and performance) due to the fact there will be a lot of messages over the session. 

Basically, my idea in regards to large files (which I still need to figure out..), is to implement some kind of InputStream and OutputStream wrappers over cometd protocol to transfer file from one server to the other without affecting the memory (in case of large files).

**Comment by sbordet** [comment]
CometD already has a server-to-server communication system, [Oort](https://docs.cometd.org/current/reference/#_java_oort). It lacks the binary support, though, since Oort is itself based on CometD.

If CometD is going to implement binary support, it's probably going to be message based, with chunks. You may then spool to disk (and/or memory) and from there use streams.

See https://docs.cometd.org/current/reference/#_support if you want to sponsor this work.

**Comment by dudicp** [comment]
Thanks Simone. I'll look into it!
